File: app/Models/Lead.php
<?php

namespace App\Models;

use Illuminate\Database\Eloquent\Model;

class Lead extends Model
{
    protected $fillable = [
        'nameEn', 'nameAr', 'email', 'phone', 'nationality', 'religion', 'gender', 'categories',
        'national_id', 'passport_no', 'birth_date', 'status', 'source', 'notes', 'parent_id', 'mother_id',
        'grade_id', 'second_language_subject_id'
    ];

    protected function casts(): array
    {
        return [
            'birth_date' => 'date',
            'categories' => 'array',
        ];
    }

    public function interactions()
    {
        return $this->hasMany(Interaction::class);
    }

    public function parent()
    {
        return $this->belongsTo(Lead::class, 'parent_id');
    }

    public function children()
    {
        return $this->hasMany(Lead::class, 'parent_id');
    }

    public function mother()
    {
        return $this->belongsTo(Lead::class, 'mother_id');
    }

    public function grade()
    {
        return $this->belongsTo(Grade::class);
    }

    public function secondLanguageSubject()
    {
        return $this->belongsTo(Subject::class, 'second_language_subject_id');
    }

    public function assignedSubjects()
    {
        if (! $this->grade) return collect();

        $subjects = $this->grade->subjectsForReligion($this->religion);

        // If the student has chosen a second language child, replace the parent
        if ($this->secondLanguageSubject) {
            $subjects = $subjects->map(function ($subject) {
                if ($subject->name === 'Second Language' && !$subject->parent_id) {
                    return $this->secondLanguageSubject;
                }
                return $subject;
            });
        }

        return $subjects->values();
    }

    public function getAgeAttribute()
    {
        if ($this->birth_date) {
            return $this->birth_date->age;
        }
        return null;
    }

    public function getDetailedAgeAttribute()
    {
        if ($this->birth_date) {
            $diff = $this->birth_date->diff(now());
            return "{$diff->y}y, {$diff->m}m, {$diff->d}d";
        }
        return null;
    }

    public static function extractBirthDateFromNationalId($nationalId)
    {
        if (strlen($nationalId) !== 14) return null;

        $century = substr($nationalId, 0, 1);
        $year = substr($nationalId, 1, 2);
        $month = substr($nationalId, 3, 2);
        $day = substr($nationalId, 5, 2);

        $fullYear = ($century == '2' ? '19' : '20') . $year;

        try {
            return \Carbon\Carbon::createFromFormat('Y-m-d', "$fullYear-$month-$day");
        } catch (\Exception $e) {
            return null;
        }
    }

    public function transferToContact(): Contact
    {
        $contact = Contact::create([
            'nameEn' => $this->nameEn,
            'nameAr' => $this->nameAr,
            'email' => $this->email,
            'phone' => $this->phone,
            'nationality' => $this->nationality,
            'religion' => $this->religion,
            'gender' => $this->gender,
            'national_id' => $this->national_id,
            'passport_no' => $this->passport_no,
            'birth_date' => $this->birth_date,
            'status' => 'Active',
            'source' => $this->source,
            'notes' => $this->notes,
            'categories' => $this->categories,
            'parent_id' => null,
            'mother_id' => null,
        ]);

        if (in_array('Student', $this->categories ?? [])) {
            Student::create([
                'contact_id' => $contact->id,
                'grade_id' => $this->grade_id ?? 1,
                'second_language_id' => $this->second_language_subject_id,
                'age_at_october' => $this->birth_date
                    ? Student::calculateAgeAtOctober($this->birth_date->format('Y-m-d'))
                    : null,
            ]);
        }

        $this->interactions()->each(function ($interaction) use ($contact) {
            $interaction->update(['contact_id' => $contact->id]);
        });

        return $contact;
    }

    protected static function booted()
    {
        static::saving(function ($lead) {
            if ($lead->nationality === 'Egyptian' && !empty($lead->national_id)) {
                $lead->birth_date = self::extractBirthDateFromNationalId($lead->national_id);
            }
        });
    }
}

File: flutter/lib/models/lead.dart
class Lead {
  final int id;
  final String nameEn;
  final String nameAr;
  final String? email;
  final String? phone;
  final String? nationality;
  final String? religion;
  final String? gender;
  final String? nationalId;
  final String? passportNo;
  final String? birthDate;
  final String status;
  final List<String> categories;
  final int? parentId;
  final int? motherId;
  final int? gradeId;
  final int? secondLanguageSubjectId;
  final String? source;
  final String? notes;
  final String? createdAt;
  final Lead? parent;
  final Lead? mother;

  Lead({
    required this.id,
    required this.nameEn,
    required this.nameAr,
    this.email,
    this.phone,
    this.nationality,
    this.religion,
    this.gender,
    this.nationalId,
    this.passportNo,
    this.birthDate,
    this.status = 'New',
    this.categories = const ['Parent'],
    this.parentId,
    this.motherId,
    this.gradeId,
    this.secondLanguageSubjectId,
    this.source,
    this.notes,
    this.createdAt,
    this.parent,
    this.mother,
  });

  factory Lead.fromJson(Map<String, dynamic> json) => Lead(
    id: json['id'],
    nameEn: json['nameEn'] ?? '',
    nameAr: json['nameAr'] ?? '',
    email: json['email'],
    phone: json['phone'],
    nationality: json['nationality'],
    religion: json['religion'],
    gender: json['gender'],
    nationalId: json['national_id'],
    passportNo: json['passport_no'],
    birthDate: json['birth_date'],
    status: json['status'] ?? 'New',
    categories: json['categories'] != null
        ? List<String>.from(json['categories'])
        : ['Parent'],
    parentId: json['parent_id'],
    motherId: json['mother_id'],
    gradeId: json['grade_id'],
    secondLanguageSubjectId: json['second_language_subject_id'],
    source: json['source'],
    notes: json['notes'],
    createdAt: json['created_at'],
    parent: json['parent'] != null ? Lead.fromJson(json['parent']) : null,
    mother: json['mother'] != null ? Lead.fromJson(json['mother']) : null,
  );

  Map<String, dynamic> toJson() => {
    'nameEn': nameEn,
    'nameAr': nameAr,
    'email': email,
    'phone': phone,
    'nationality': nationality,
    'religion': religion,
    'gender': gender,
    'national_id': nationalId,
    'passport_no': passportNo,
    'birth_date': birthDate,
    'status': status,
    'categories': categories,
    'parent_id': parentId,
    'mother_id': motherId,
    'grade_id': gradeId,
    'second_language_subject_id': secondLanguageSubjectId,
    'source': source,
    'notes': notes,
  };
}
